File: algorithm1/binary_search/ts/FirstBadVersion.ts
/**
 * The knows API is defined in the parent class Relation.
 * isBadVersion(version: number): boolean {
 *     ...
 * };
 */

var solution = function(isBadVersion: any) {
  return function(n: number): number {
    let left = 0, right = n;

    while (left < right) {
      const middle = left + Math.floor((right - left)/2); // 중간지점

      if (isBadVersion(middle)) {
        right = middle;
        // minimum 체크하고 해당 지점 이전 탐색 
      } else {
        // 해당 지점 이후로 탐색
        left = middle + 1;
      }
      
    }
    return left;
  };  
};

File: algorithm1/binary_search/python/first_bad_version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:
import math

class Solution:
    def firstBadVersion(self, n: int) -> int:
      left = 0
      right = n

      while left < right :
        middle = left + math.floor((right - left) / 2)
        
        if isBadVersion(middle):
          right = middle
        else:
          left = middle + 1
      return left